internal/check/engine.go

```diff
@@ -35,6 +39,11 @@ func (e *Engine) subjectIsAllowed(ctx context.Context, requested *relationtuple.
 	// TODO replace by more performant algorithm: https://github.com/ory/keto/issues/483
 
 	for _, sr := range rels {
+		ctx, wasAlreadyVisited := e.u.CheckVisited(ctx, sr.String())
```

Comment:
Here you add the whole relation tuple as a string. Maybe it would be better to stay consistent and only record visited subject sets?